### ISSUE TITLE
[Week 4] [Critical]  Inverted PnL Calculation Causes Total Sol Mismanagement in WithdrawFromYield

### ISSUE THREAD

**Issue by syed-ghufran-hassan** [Frankcastleauditor/Solana-Audit-Arena#175]
Week: 4
Researcher: GitHub @syed-ghufran-hassan + X @SyedGhufranHas1
Severity: Critical
Category: Logic Error / Accounting Bug
Affected function(s): FrankSol/programs/stake_v2/src/instructions/withdraw_from_yield.rs::handler

## Description

The withdraw_from_yield handler in `withdraw_from_yield.rs` contains a critical accounting error where Profit and Loss (PnL) updates to pool.total_sol are applied in the opposite direction of what economic logic requires.

When withdrawing funds from an external yield program, the function incorrectly handles the PnL calculation:

```rust
if pnl_i128 >= 0 {
    // PROFIT case - INCORRECT: decreases total_sol
    let gain = u64::try_from(pnl_i128)?;
    pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), gain)?);
} else {
    // LOSS case - INCORRECT: increases total_sol
    let loss = u64::try_from(-pnl_i128)?;
    pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), loss)?);
}
```

The logic is backwards:

- When the protocol earns a profit → total_sol should INCREASE (more SOL backing each frankSOL)

- When the protocol suffers a loss → total_sol should DECREASE (less SOL backing each frankSOL)

- Instead, the current implementation does the opposite, which would severely damage the protocol's economics.

### Scenario 1: Profitable Strategy (e.g., +10 SOL profit)

- What should happen: total_sol increases by 10 SOL → frankSOL appreciates in value

- What actually happens: total_sol decreases by 10 SOL → frankSOL loses value despite profit

- Result: Stakers are penalized when the protocol succeeds

### Scenario 2: Losing Strategy (e.g., -10 SOL loss)

- What should happen: total_sol decreases by 10 SOL → frankSOL depreciates

- What actually happens: total_sol increases by 10 SOL → frankSOL gains value despite loss

- Result: Stakers are rewarded when the protocol loses money.

##  POC

```rust
- Fund manager deploys 100 SOL to a yield strategy

- Yield strategy generates 10 SOL profit

- Upon calling withdraw_from_yield: Protocol incorrectly records a 10 SOL loss

- Exchange rate incorrectly drops, hurting all stakers

- Malicious fund manager could repeat this to systematically drain value from frankSOL holders
```

```rust
#[test]
fn test_pnl_accounting_bug_in_withdraw_from_yield() {
    // Initial state
    let mut pool = Pool {
        total_sol: PodU64::from(1000),
        deployed_sol: PodU64::from(200),
        franksol_supply: PodU64::from(1000),
        ..Default::default()
    };
    
    // Simulate profit scenario from yield withdrawal
    let vault_before = 800;
    let principal_returned = 100;
    let vault_after = 905; // Received 105 total (100 principal + 5 profit)
    let total_return = 105;
    let pnl = 5; // PROFIT of 5 SOL
    
    // BUGGY BEHAVIOR in withdraw_from_yield
    let gain = pnl;
    pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), gain).unwrap());
    // total_sol becomes: 1000 - 5 = 995 SOL
    
    // WHAT SHOULD HAPPEN
    // total_sol should be: 1000 + 5 = 1005 SOL
    
    // Assert the bug
    assert_eq!(pool.total_sol.get(), 995); // Bug: should be 1005
    // Result: 10 SOL value discrepancy
}
```

## Fix 

Correct Implementation for withdraw_from_yield.rs:

```rust
pub fn handler(
    ctx: &mut Context<WithdrawFromYield>,
    principal_returned: u64,
    yield_amount: u64,
) -> Result<()> {
    let pool = &mut ctx.accounts.pool;
    require_keys_eq!(
        *ctx.accounts.fund_manager.address(),
        pool.fund_manager,
        StakeError::Unauthorized
    );
    require!(principal_returned > 0, StakeError::InvalidAmount);
    require!(
        principal_returned <= pool.deployed_sol.get(),
        StakeError::InsufficientDeployed
    );
    
    let vault_before = ctx.accounts.vault.lamports();
    
    // CPI to withdraw from yield program
    let cpi_accounts = yield_generator::cpi::accounts::Withdraw {
        operator: ctx.accounts.fund_manager.cpi_handle(),
        state: ctx.accounts.yield_state.cpi_handle_mut(),
        position: ctx.accounts.yield_position.cpi_handle_mut(),
        yield_vault: ctx.accounts.yield_vault.cpi_handle_mut(),
        destination_vault: ctx.accounts.vault.cpi_handle_mut(),
        system_program: ctx.accounts.system_program.cpi_handle(),
    };
    let cpi_ctx = CpiContext::new(ctx.accounts.yield_generator_program.address(), cpi_accounts);
    yield_generator::cpi::withdraw(cpi_ctx, principal_returned, yield_amount)?;
    
    let vault_after = ctx.accounts.vault.lamports();
    let total_return = checked_sub_u64(vault_after, vault_before)?;
    let pnl_i128 = (total_return as i128) - (principal_returned as i128);
    
    // Update deployed SOL (always decreases by principal)
    pool.deployed_sol = PodU64::from(checked_sub_u64(
        pool.deployed_sol.get(),
        principal_returned,
    )?);
    
    // CORRECT PnL accounting
    if pnl_i128 >= 0 {
        // PROFIT: total_sol INCREASES
        let gain = u64::try_from(pnl_i128).map_err(|_| StakeError::MathOverflow)?;
        pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), gain)?);
    } else {
        // LOSS: total_sol DECREASES
        let loss = u64::try_from(-pnl_i128).map_err(|_| StakeError::MathOverflow)?;
        pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), loss)?);
    }
    
    Ok(())
}
```

**Comment by novoyd** [comment]
This is the same inverted-PnL root already covered by #144: `withdraw_from_yield` applies gains/losses in the wrong direction when updating `pool.total_sol`. Since #144 was submitted first, I’d treat this as duplicate.

**Comment by Frankcastleauditor** [comment]
✅ Great Valid but Duplicate finding
